### PR TITLE
ARROW-10821 [Rust][Datafusion] support negative expression

### PR DESCRIPTION
**Pull request by houqp**
To support queries like `SELECT c3 FROM aggregate_test_100 WHERE -c4 > 0`:

  * add negate compute kernel in arrow
  * add negative expression in datafusion
  * support negative and positive operators in datafusion's sql planner

### CODE REVIEW

File: rust/arrow/src/array/cast.rs [hunk 1]
```diff
@@ -59,5 +59,6 @@ macro_rules! array_downcast_fn {
 }
 
 array_downcast_fn!(as_string_array, StringArray);
+array_downcast_fn!(as_largestring_array, LargeStringArray);
```
Comment (houqp) [inline rust/arrow/src/array/cast.rs:62]: drive by change.

**Review by jorgecarleitao — APPROVED** [review]
LGTM. Thanks @houqp

File: rust/datafusion/src/physical_plan/expressions.rs [hunk 2]
```diff
@@ -1682,6 +1691,82 @@ pub fn not(
...
+            ColumnarValue::Array(array) => {
+                let result: Result<ArrayRef> = match array.data_type() {
+                    DataType::Int8 => compute_op!(array, negate, Int8Array),
+                    DataType::Int16 => compute_op!(array, negate, Int16Array),
+                    DataType::Int32 => compute_op!(array, negate, Int32Array),
+                    DataType::Int64 => compute_op!(array, negate, Int64Array),
+                    DataType::Float32 => compute_op!(array, negate, Float32Array),
+                    DataType::Float64 => compute_op!(array, negate, Float64Array),
+                    _ => panic!(
+                        "(- '{:?}') can't be evaluated because the expression's type is {:?}, not signed numeric",
```
[the remark this answers is not in the document]
Comment (houqp) [reply rust/datafusion/src/physical_plan/expressions.rs:1734]: good point, let me make the change

[the remark this answers is not in the document]
Comment (houqp) [reply rust/datafusion/src/physical_plan/expressions.rs:1734]: pushed a commit to change this to internal error. also replaced the unwrap in literal parsing with internal error as well.

**Pull request closed without merging**